File: guides/managed-agents/cursor-cloud-agents/typescript/spawn.mjs
#!/usr/bin/env node
// --spawn hook for `agent worker controller`. Runs once per claimed request
// (or once per missing warm worker) with CURSOR_* set by the controller.
// Creates a Superserve sandbox and starts a Cursor pool worker inside it.
import "./env.mjs"
import { Sandbox } from "@superserve/sdk"

import {
  META_MANAGED,
  META_LAUNCHING,
  META_POOL,
  META_REPO,
  META_REQUEST_ID,
  META_WORKER_ID,
  config,
  findSandboxForWorker,
  launchWorker,
  releaseClaim,
  tagSandbox,
  workerEnv,
  workerState,
} from "./worker.mjs"

for (const key of [
  "SUPERSERVE_API_KEY",
  "CURSOR_API_KEY",
  "CURSOR_AGENT_WORKER_ID",
  "CURSOR_POOL",
]) {
  if (!process.env[key]) {
    console.error(`spawn: ${key} is not set`)
    process.exit(2)
  }
}

const workerId = process.env.CURSOR_AGENT_WORKER_ID
const pool = process.env.CURSOR_POOL
const requestId = process.env.CURSOR_REQUEST_ID
const workerName = process.env.CURSOR_WORKER_NAME
const repo =
  process.env.CURSOR_REPO_OWNER && process.env.CURSOR_REPO_NAME
    ? `${process.env.CURSOR_REPO_OWNER}/${process.env.CURSOR_REPO_NAME}`
    : undefined

function sandboxName() {
  const short = workerId
    .toLowerCase()
    .replace(/[^a-z0-9]/g, "")
    .slice(0, 12)
  return `cursor-${short || "worker"}`
}

async function createSandbox() {
  const metadata = {
    [META_MANAGED]: "true",
    [META_WORKER_ID]: workerId,
    [META_POOL]: pool,
  }
  if (requestId) metadata[META_REQUEST_ID] = requestId
  if (repo) metadata[META_REPO] = repo

  const options = {
    name: sandboxName(),
    fromTemplate: config.templateName,
    metadata,
    autoDeleteSeconds: config.autoDeleteSeconds,
  }
  if (config.allowOut.length > 0) {
    options.network = { allowOut: config.allowOut, denyOut: ["0.0.0.0/0"] }
  }
  return Sandbox.create(options)
}

// Hand the request back and drop a sandbox we created, so a failed spawn
// never strands a claim or leaves an idle sandbox running.
async function abandon(sandbox, created) {
  if (requestId) {
    try {
      await releaseClaim(requestId)
      console.error(`spawn: released claim request=${requestId}`)
    } catch (e) {
      console.error(
        `spawn: could not release claim request=${requestId}: ${e.message}`,
      )
    }
  }
  if (created && sandbox) {
    try {
      await sandbox.kill()
    } catch (e) {
      console.error(
        `spawn: could not delete sandbox=${sandbox.id}: ${e.message}`,
      )
    }
  }
}

async function main() {
  let sandbox
  let created = false

  try {
    // A paused sandbox tagged with this worker id is a hibernated workspace
    // (see the monitor). Resume it instead of starting from scratch.
    const existing = await findSandboxForWorker(workerId)
    if (existing) {
      console.log(
        `spawn: reusing sandbox=${existing.id} status=${existing.status} worker=${workerId}`,
      )
      // Tell the monitor a relaunch is in progress before anything resumes the
      // sandbox: connect() auto-resumes, and the previous worker's exit file
      // is visible the moment it does.
      await Sandbox.updateById(existing.id, {
        metadata: {
          ...existing.metadata,
          [META_LAUNCHING]: String(Date.now()),
        },
      })
      sandbox = await Sandbox.connect(existing.id)
      if (sandbox.status === "paused") await sandbox.resume()
      const state = await workerState(sandbox)
      if (state.state === "running") {
        console.log(
          `spawn: worker already running pid=${state.pid} sandbox=${sandbox.id}`,
        )
        return 0
      }
    } else {
      sandbox = await createSandbox()
      created = true
      console.log(
        `spawn: created sandbox=${sandbox.id} template=${config.templateName} worker=${workerId}`,
      )
    }

    const result = await launchWorker(sandbox, {
      pool,
      env: workerEnv({ workerId, workerName }),
    })
    if (!created) await tagSandbox(sandbox, { [META_LAUNCHING]: null })
    if (!result.ok) {
      console.error(
        `spawn: worker failed to start sandbox=${sandbox.id} state=${result.state.state} ` +
          `exit=${result.state.exit_code ?? "-"}\n${result.log.slice(-2000)}`,
      )
      await abandon(sandbox, created)
      return 1
    }

    console.log(
      `spawn: worker started pid=${result.pid} sandbox=${sandbox.id} pool=${pool} worker=${workerId}` +
        (requestId ? ` request=${requestId}` : ""),
    )
    return 0
  } catch (e) {
    console.error(`spawn: ${e.message}`)
    await abandon(sandbox, created)
    return 1
  }
}

process.exit(await main())

File: guides/managed-agents/cursor-cloud-agents/python/spawn.py
"""--spawn hook for `agent worker controller`.

Runs once per claimed request (or once per missing warm worker) with CURSOR_*
set by the controller. Creates a Superserve sandbox and starts a Cursor pool
worker inside it.
"""

from __future__ import annotations

import os
import re
import sys
import time

from superserve import NetworkConfig, Sandbox
from worker import (
    ALLOW_OUT,
    AUTO_DELETE_SECONDS,
    META_LAUNCHING,
    META_MANAGED,
    META_POOL,
    META_REPO,
    META_REQUEST_ID,
    META_WORKER_ID,
    TEMPLATE_NAME,
    find_sandbox_for_worker,
    launch_worker,
    release_claim,
    status_of,
    tag_sandbox,
    worker_env,
    worker_state,
)


def log(message: str) -> None:
    print(f"spawn: {message}", flush=True)


def err(message: str) -> None:
    print(f"spawn: {message}", file=sys.stderr, flush=True)


def sandbox_name(worker_id: str) -> str:
    short = re.sub(r"[^a-z0-9]", "", worker_id.lower())[:12]
    return f"cursor-{short or 'worker'}"


def create_sandbox(
    worker_id: str, pool: str, request_id: str | None, repo: str | None
) -> Sandbox:
    metadata = {META_MANAGED: "true", META_WORKER_ID: worker_id, META_POOL: pool}
    if request_id:
        metadata[META_REQUEST_ID] = request_id
    if repo:
        metadata[META_REPO] = repo

    network = None
    if ALLOW_OUT:
        network = NetworkConfig(allow_out=ALLOW_OUT, deny_out=["0.0.0.0/0"])

    return Sandbox.create(
        name=sandbox_name(worker_id),
        from_template=TEMPLATE_NAME,
        metadata=metadata,
        auto_delete_seconds=AUTO_DELETE_SECONDS,
        network=network,
    )


def abandon(sandbox: Sandbox | None, created: bool, request_id: str | None) -> None:
    """Hand the request back and drop a sandbox we created.

    A failed spawn must never strand a claim or leave an idle sandbox running.
    """
    if request_id:
        try:
            release_claim(request_id)
            err(f"released claim request={request_id}")
        except Exception as e:
            err(f"could not release claim request={request_id}: {e}")
    if created and sandbox is not None:
        try:
            sandbox.kill()
        except Exception as e:
            err(f"could not delete sandbox={sandbox.id}: {e}")


def main() -> int:
    for key in (
        "SUPERSERVE_API_KEY",
        "CURSOR_API_KEY",
        "CURSOR_AGENT_WORKER_ID",
        "CURSOR_POOL",
    ):
        if not os.environ.get(key):
            err(f"{key} is not set")
            return 2

    worker_id = os.environ["CURSOR_AGENT_WORKER_ID"]
    pool = os.environ["CURSOR_POOL"]
    request_id = os.environ.get("CURSOR_REQUEST_ID") or None
    worker_name = os.environ.get("CURSOR_WORKER_NAME") or None
    owner, name = (
        os.environ.get("CURSOR_REPO_OWNER"),
        os.environ.get("CURSOR_REPO_NAME"),
    )
    repo = f"{owner}/{name}" if owner and name else None

    sandbox: Sandbox | None = None
    created = False
    try:
        # A paused sandbox tagged with this worker id is a hibernated workspace
        # (see the monitor). Resume it instead of starting from scratch.
        existing = find_sandbox_for_worker(worker_id)
        if existing:
            log(
                f"reusing sandbox={existing.id} status={status_of(existing)} worker={worker_id}"
            )
            # Tell the monitor a relaunch is in progress before anything resumes
            # the sandbox: connect() auto-resumes, and the previous worker's exit
            # file is visible the moment it does.
            Sandbox.update_by_id(
                existing.id,
                metadata={
                    **(existing.metadata or {}),
                    META_LAUNCHING: str(int(time.time() * 1000)),
                },
            )
            sandbox = Sandbox.connect(existing.id)
            if status_of(sandbox) == "paused":
                sandbox.resume()
            state = worker_state(sandbox)
            if state["state"] == "running":
                tag_sandbox(sandbox, {META_LAUNCHING: None})
                log(f"worker already running pid={state['pid']} sandbox={sandbox.id}")
                return 0
        else:
            sandbox = create_sandbox(worker_id, pool, request_id, repo)
            created = True
            log(
                f"created sandbox={sandbox.id} template={TEMPLATE_NAME} worker={worker_id}"
            )

        result = launch_worker(sandbox, pool, worker_env(worker_id, worker_name))
        if not created:
            tag_sandbox(sandbox, {META_LAUNCHING: None})
        if not result["ok"]:
            state = result["state"]
            err(
                f"worker failed to start sandbox={sandbox.id} state={state.get('state')} "
                f"exit={state.get('exit_code', '-')}\n{result['log'][-2000:]}"
            )
            abandon(sandbox, created, request_id)
            return 1

        suffix = f" request={request_id}" if request_id else ""
        log(
            f"worker started pid={result['pid']} sandbox={sandbox.id} "
            f"pool={pool} worker={worker_id}{suffix}"
        )
        return 0
    except Exception as e:
        err(str(e))
        abandon(sandbox, created, request_id)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
